File: Data Structures & Algorithms/partition-equal-subset-sum/submission-9.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        s = sum(nums)
        if s % 2:
            return False
        target = s // 2

        dp = {0}
        for num in nums:
            dp_tmp = set()
            for current_sum in dp:
                if current_sum == target:
                    return True
                dp_tmp |= {current_sum, current_sum + num}
            dp = dp_tmp

        return False

File: Data Structures & Algorithms/two-integer-sum/submission-9.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> m;
        for (int i = 0; i < nums.size(); ++i) {
            if (m.contains(nums[i])) {
                return vector<int>{m[nums[i]], i};
            }
            m[target - nums[i]] = i;
        }

        return vector<int>{-1, -1};
    }
};

